### ISSUE TITLE
Design input for model state

### ISSUE THREAD

**Issue by JoyMonteiro** [chapel-lang/chapel#17593]
Hello,

we are trying to design a data structure which would function as model state for a climate model.

We have an implementation in Python currently (https://sympl.readthedocs.io), but would like to shift to Chapel to take advantage of multi-locale execution.

The requirement is as follows:
* Each model component takes a model state which is currently a dictionary with strings as keys and a labeled array container (DataArray) as value
* DataArrays contain an array of arbitrary type (we could restrict it to int, real and string)
* The component extracts arrays that are relevant to it (not all may be relevant) and uses these arrays to perform some computations
* Components behave like functions which take model state as input and provide a similar state variable with tendencies  or new model state

In Python, the model state is simply a dictionary and it works fine. However, a key-value store does not seem the ideal approach in Chapel because of typing issues (DataArray is a generic).

Just to give you an idea of the current design and how it is used, `state` is the model state here:
https://github.com/CliMT/climt/blob/2f03b11e2c44f454ca3f353d61c0ad9ffc27fcd2/examples/ice_sheet.py#L8

Is there some Chapel feature/ibrary that we are missing that could make this easy for us?

TIA,
Joy

**Comment by bradcray** [comment]
Hi Joy —

[Meta: Were you asking about this on gitter earlier this week?  I saw some messages asking about Maps with varying value types, but the gitter handles from which they were posted were hard to parse, and seemed to change between questions.  As a meta-comment, it might be good to move this query and discussion to the [Users category of the Chapel Discourse site](https://chapel.discourse.group/c/users/6), as it seems more likely to be an ongoing discussion than an issue for our team to address and close, like a bug or feature request; and it's deep enough to be too heavyweight for gitter].

One approach you could take here would be to have a shallow class hierarchy to take care of the generic nature of the DataArrays, where the map would store the type-less abstract parent class as its value type.  For example, imagine a pattern like:

```chapel
class AbstractDataArray {
}

class DataArray: AbstractDataArray {
  type eltType;
  const n: int;
  var A: [1..n] eltType;
}

var map: Map(string, AbstractDataArray);
```

The challenge to doing this is that when you do a lookup in your map, you'll get back a very generic thing, and will have to determine what type it is before you can do anything particularly useful with it.  For example, you could try to dynamically cast it to each of the possible `DataArray` types until you figure out which one it is; or you could have the parent class store an enum or other indicator of which "flavor" of DataArray it actually is, and then use that to determine which type to dynamically cast it to.  I don't know whether you're familiar with the [Arkouda project](https://github.com/mhmerrill/arkouda), but this is the approach they take in order to get a similar type of multi-type table, where the abstract class is [GenSymEntry](https://github.com/mhmerrill/arkouda/blob/master/src/MultiTypeSymEntry.chpl#L27), the child class is [SymEntry](https://github.com/mhmerrill/arkouda/blob/master/src/MultiTypeSymEntry.chpl#L58), and [SymTab](https://github.com/mhmerrill/arkouda/blob/master/src/MultiTypeSymbolTable.chpl#L16) is their class that stores the `map`.

Of course, you could also create methods on the AbstractDataArray and DataArray classes and rely on dynamic dispatch to implement the methods without returning the concrete types, and this will work for some things; but once you need the method to return a value of `eltType` or to interact with another `DataArray`, this tends to get more complicated due to the constraints of dynamic dispatch; but for some things (like querying the value of "n"), it could be a good way to go.

I didn't get to this until late in the day, but wanted to get some thoughts back to you before the weekend, so am going to pause here for now and see whether this generally makes sense or whether you have further questions / need more information to see a path forward.

**Comment by JoyMonteiro** [comment]
Thanks, Brad. I think the Gitter query was by some of my colleagues. But I agree this is more suited for Discourse, I'll move it there.
